File: android/src/main/java/expo/modules/logisticsscanner/scanner/ScannerManager.kt
package expo.modules.logisticsscanner.scanner

import android.content.Context
import android.graphics.SurfaceTexture
import androidx.camera.view.PreviewView
import androidx.lifecycle.LifecycleOwner

class ScannerManager(
  private val context: Context,
  private val lifecycleOwner: LifecycleOwner,
  private val previewView: PreviewView,
) {
  private val provider: CameraScannerProvider = CameraScannerProvider(
    context,
    lifecycleOwner,
    previewView,
  )

  fun start(onScan: (ScanResult) -> Unit) {
    start(onScan, onStarted = {}, onFailed = {}, onError = {})
  }

  fun start(
    onScan: (ScanResult) -> Unit,
    onStarted: () -> Unit,
    onFailed: (ScanError) -> Unit,
    onError: (ScanError) -> Unit,
  ) {
    provider.start(onScan, onStarted, onFailed, onError)
  }

  fun stop() {
    provider.stop()
  }

  fun setTorch(enabled: Boolean) {
    provider.setTorch(enabled)
  }

  fun setPreviewSurfaceTexture(texture: SurfaceTexture?) {
    provider.setPreviewSurfaceTexture(texture)
  }
}

File: ios/scanner/ScannerManager.swift
import UIKit

final class ScannerManager {
  private let previewView: UIView
  private let provider: CameraScannerProvider

  init(previewView: UIView) {
    self.previewView = previewView
    self.provider = CameraScannerProvider(previewView: previewView)
  }

  func start(
    onScan: @escaping (ScanResult) -> Void,
    onStarted: @escaping () -> Void = {},
    onFailed: @escaping () -> Void = {}
  ) {
    provider.start(onScan: onScan, onStarted: onStarted, onFailed: onFailed)
  }

  func stop() {
    provider.stop()
  }

  func setTorch(enabled: Bool) {
    provider.setTorch(enabled: enabled)
  }

  func updatePreviewFrame(_ frame: CGRect) {
    provider.updatePreviewFrame(frame)
  }
}
